source/server/listener_manager_impl.h

```diff
@@ -305,36 +309,67 @@ class ListenerImpl : public Network::ListenerConfig,
+  void addFilterChainForDestinationIPs(DestinationIPsMap& destination_ips_map,
+                                       const std::vector<std::string>& destination_ips,
+                                       const std::vector<std::string>& server_names,
+                                       const std::string& transport_protocol,
+                                       const std::vector<std::string>& application_protocols,
+                                       const Network::FilterChainSharedPtr& filter_chain);
+  void addFilterChainForServerNames(ServerNamesMap& server_names_map,
+                                    const std::vector<std::string>& server_names,
+                                    const std::string& transport_protocol,
+                                    const std::vector<std::string>& application_protocols,
+                                    const Network::FilterChainSharedPtr& filter_chain);
+  void addFilterChainForApplicationProtocols(ApplicationProtocolsMap& application_protocol_map,
+                                             const std::vector<std::string>& application_protocols,
+                                             const Network::FilterChainSharedPtr& filter_chain);
+
+  void finishFilterChain();
+
+  const Network::FilterChain*
+  findFilterChainForDestinationIP(const DestinationIPsTrie& destination_ips_trie,
+                                  const Network::ConnectionSocket& socket) const;
+  const Network::FilterChain*
+  findFilterChainForServerName(const ServerNamesMap& server_names_map,
+                               const Network::ConnectionSocket& socket) const;
+  const Network::FilterChain*
+  findFilterChainForTransportProtocol(const TransportProtocolsMap& transport_protocols_map,
+                                      const Network::ConnectionSocket& socket) const;
+  const Network::FilterChain*
+  findFilterChainForApplicationProtocols(const ApplicationProtocolsMap& application_protocols_map,
+                                         const Network::ConnectionSocket& socket) const;
+
   static bool isWildcardServerName(const std::string& name);
 
-  // Mapping of FilterChain's configured server name and transport protocol, i.e.
-  //   map[server_name][transport_protocol][application_protocol] => FilterChainSharedPtr
-  //
-  // For the server_name lookups, both exact server names and wildcard domains are part of the same
-  // map, in which wildcard domains are prefixed with "." (i.e. ".example.com" for "*.example.com")
-  // to differentiate between exact and wildcard entries.
-  std::unordered_map<
-      std::string, std::unordered_map<
-                       std::string, std::unordered_map<std::string, Network::FilterChainSharedPtr>>>
-      filter_chains_;
+  // Mapping of FilterChain's configured destination ports, IPs, server names, transport protocols
+  // and application protocols, using structures defined above.
+  DestinationPortsMap filter_chains_;
```

Comment:
Maybe rename to destination_ports_map_? I both like the current name, and dislike it. From a high level, filter_chains_ feels like the right name, but when reading the functions that use it, it is hard to follow. So, I'm not sure what I'm suggesting with this comment.

Reply:
As I commented above, the code would read better if we simply called this destination_ports_map (in the filter chain match section, which is the crux of this entire thing)

Reply:
Done.